evm/forge/tests/CircleIntegration.t.sol

```diff
@@ -0,0 +1,491 @@
+
+        (uint16 emitterChain, uint32 remoteDomain) = _registerEmitterAndDomain();
+
+        ICircleIntegration.RedeemParameters memory redeemParams = circleIntegration
+            .craftRedeemParameters(
+            CraftedCctpMessageParams({
+                remoteDomain: remoteDomain,
+                nonce: 2 ** 64 - 1,
+                remoteToken: remoteToken,
+                mintRecipient: address(this).toUniversalAddress(),
+                amount: 69
+            }),
+            CraftedVaaParams({emitterChain: emitterChain, sequence: 88}),
+            0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef, // fromAddress
+            abi.encodePacked("Somebody set up us the bomb")
+        );
+
+        // You shall not pass!
+        _expectRevert(
+            abi.encodeCall(circleIntegration.redeemTokensWithPayload, (redeemParams)),
+            abi.encodeCall(this.Error, ("Mint token not supported")) // CCTP Token Minter error
+        );
+    }
+
+    function test_CannotRedeemTokensWithPayloadInvalidMessagePair() public {
+        (uint16 emitterChain, uint32 remoteDomain) = _registerEmitterAndDomain();
+
+        ICircleIntegration.RedeemParameters memory redeemParams1 = circleIntegration
+            .craftRedeemParameters(
+            CraftedCctpMessageParams({
+                remoteDomain: remoteDomain,
+                nonce: 2 ** 64 - 2,
+                remoteToken: FOREIGN_USDC_ADDRESS,
+                mintRecipient: address(this).toUniversalAddress(),
+                amount: 69
+            }),
+            CraftedVaaParams({emitterChain: emitterChain, sequence: 88}),
+            0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef, // fromAddress
+            abi.encodePacked("Somebody set up us the bomb")
+        );
+
+        ICircleIntegration.RedeemParameters memory redeemParams2 = circleIntegration
+            .craftRedeemParameters(
+            CraftedCctpMessageParams({
+                remoteDomain: remoteDomain,
+                nonce: 2 ** 64 - 1,
+                remoteToken: FOREIGN_USDC_ADDRESS,
+                mintRecipient: address(this).toUniversalAddress(),
+                amount: 69
+            }),
+            CraftedVaaParams({emitterChain: emitterChain, sequence: 89}),
+            0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef, // fromAddress
+            abi.encodePacked("Somebody set up us the bomb")
+        );
+
+        // Swap VAAs.
+        bytes memory tmpEncodedVaa = redeemParams2.encodedVaa;
+        redeemParams2.encodedVaa = redeemParams1.encodedVaa;
+        redeemParams1.encodedVaa = tmpEncodedVaa;
+
+        // You shall not pass!
+        _expectRevert(
+            abi.encodeCall(circleIntegration.redeemTokensWithPayload, (redeemParams1)),
+            abi.encodeCall(this.Error, ("invalid message pair"))
+        );
+    }
+
+    function test_RedeemTokensWithPayload() public {
```

Comment:
Another nit - we should do a balance check in this test.

Reply:
Good catch, fixed the test here: https://github.com/wormhole-foundation/wormhole-circle-integration/pull/48/commits/15aa47c70ab413ff7553753ee7acd5bf3f1277a6